### PR TITLE
nixos/cc-wrapper: Fix bug if dynamicLinker not found

### PR DESCRIPTION
**Pull request by bnikolic**
If a dynamic linker for target is not found the generated script fails
due to unbound variable error (in turn due to "set -u"). Correct by specifying
default value with dynamicLinker:- and not generating ldflagsBefore if
no linker is found.

This problem was found when cross compiling to mingw32 targets

###### Motivation for this change

Cross-compiling for mingw32 targets was currently broken with "variable unbound" errors. This change allows the compilation to proceed further.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vcunat** [comment]
@Ericson2314 knows cross-compiling best, I guess.

**Comment by orivej** [comment]
This does not look right, the variable is always initialized a few lines above. And why would `dynamicLinker=($dynamicLinker)` not fail too?

**Comment by bnikolic** [comment]
Because

dynamicLinker=()

in bash unbinds the variable.

(Try it! E.g.

  set -u
  vv="dkj"
  vv=()
  echo $vv
    
fails with unbound variable)

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -286,7 +286,7 @@ stdenv.mkDerivation {
         *) echo "Multiple dynamic linkers found for platform '${targetPlatform.config}'." >&2;;
       esac
 
-      if [ -n "$dynamicLinker" ]; then
+      if [ -n "''${dynamicLinker:-}" ]; then
```
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: This is correct, good catch.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 2]
```diff
@@ -296,7 +296,9 @@ stdenv.mkDerivation {
           echo ${libc_lib}/lib/32/ld-linux.so.2 > $out/nix-support/dynamic-linker-m32
         fi
 
-        local ldflagsBefore=(-dynamic-linker "$dynamicLinker")
+     if [ -n "''${dynamicLinker:-}" ]; then
```
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: This second conditional is unneeded, no?

Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix]: It's already part of the first, unless I messed up the string interpolation.

Comment (bnikolic) [reply pkgs/build-support/cc-wrapper/default.nix]: You're right, I missed the indentation. Pushing a revised patch now

**Review by orivej — APPROVED** [review]

**Comment by globin** [comment]
@nixborg build

**Review by vcunat — APPROVED** [review]

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31209

**Comment by vcunat** [comment]
@globin: it's included in staging now, so you may want to save resources for other stuff.

**Comment by Ericson2314** [comment]
Thanks @bnikolic! Btw I'm happy to pitch on various MinGW-related things, especially as they bitrotted further a bit under my watch. @domenkozar and @DavidEGrayson may also be interested.

**Comment by bnikolic** [comment]
Thanks. Things are now failing a bit further along, looks like a problem
between sys-includes and includes when building mingw32 (sys-includes
need to come first by they do not by default?). I'll try to have a look
at it later in the week but if you one of you already has some ideas why
this could be happening do let me know...

**Comment by DavidEGrayson** [comment]
@bnikolic  I'm not sure if this is helpful for your sys-includes problem, but I use two patches for mingw-w64 GCC to make the search paths correct:

https://github.com/DavidEGrayson/nixcrpkgs/blob/ 6501664/mingw-w64/gcc/cppdefault.patch
https://github.com/DavidEGrayson/nixcrpkgs/blob/ 6501664/mingw-w64/gcc/mingw-search-paths.patch

You might also want to look at the Arch Linux recipe for mingw-w64-gcc:

https://www.archlinux.org/packages/community/x86_64/mingw-w64-gcc/

**Comment by bnikolic** [comment]
@DavidEGrayson I've resolved the mingw-w64 problems in #33167 -- cross-compilation to win64 target seems to work with this.

**Pull request merged**